Add __size_returning_new variant detection to TLI.

snehasish opened the pull request:
Add support to detect __size_returning_new variants defined inproposal P0901R5 to extend to operator new, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0901r5.html for details.

This PR matches the declarations exported by tcmalloc in https://github.com/google/tcmalloc/blob/f2516691d01051defc558679f37720bba88d9862/tcmalloc/malloc_extension.h#L707-L711

Commit:
Add __size_returning_new variant detection to TLI.

Add support to detect __size_returning_new variants defined inproposal P0901R5 to extend to
operator new, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p0901r5.html for
details.

This PR matches the declarations exported by tcmalloc in
https://github.com/google/tcmalloc/blob/f2516691d01051defc558679f37720bba88d9862/tcmalloc/malloc_extension.h#L707-L711

teresajohnson reviewed (approved):
lgtm

nikic reviewed (commented):
Missing validation of the return struct type?

snehasish commented:
Added a unittest to validate the return type (also updated up the return types in the test) in https://github.com/llvm/llvm-project/pull/102391. @nikic ptal, thanks!
